superset/dashboards/permalink/api.py

```diff
@@ -0,0 +1,174 @@
+from superset.dashboards.permalink.exceptions import DashboardPermalinkInvalidStateError
+from superset.dashboards.permalink.schemas import (
+    DashboardPermalinkPostSchema,
+    DashboardPermalinkPutSchema,
+)
+from superset.extensions import event_logger
+from superset.key_value.exceptions import KeyValueAccessDeniedError
+from superset.views.base_api import requires_json
+
+logger = logging.getLogger(__name__)
+
+
+class DashboardPermalinkRestApi(BaseApi):
+    add_model_schema = DashboardPermalinkPostSchema()
+    edit_model_schema = DashboardPermalinkPutSchema()
+    method_permission_name = MODEL_API_RW_METHOD_PERMISSION_MAP
+    include_route_methods = {
+        RouteMethod.POST,
+        RouteMethod.PUT,
+        RouteMethod.GET,
+        RouteMethod.DELETE,
+    }
+    allow_browser_login = True
+    class_permission_name = "DashboardPermalinkRestApi"
+    resource_name = "dashboard"
+    openapi_spec_tag = "Dashboard Permanent Link"
+    openapi_spec_component_schemas = (
+        DashboardPermalinkPostSchema,
+        DashboardPermalinkPutSchema,
+    )
+
+    @expose("/<pk>/permalink", methods=["POST"])
+    @protect()
+    @safe
+    @event_logger.log_this_with_context(
+        action=lambda self, *args, **kwargs: f"{self.__class__.__name__}.post",
+        log_to_statsd=False,
+    )
+    @requires_json
+    def post(self, pk: str) -> Response:
+        """Stores a new permanent link.
+        ---
+        post:
+          description: >-
+            Stores a new permanent link.
+          parameters:
+          - in: path
+            schema:
+              type: string
+            name: pk
+          requestBody:
+            required: true
+            content:
+              application/json:
+                schema:
+                  $ref: '#/components/schemas/DashboardPermalinkPostSchema'
+          responses:
+            201:
+              description: The permanent link was stored successfully.
+              content:
+                application/json:
+                  schema:
+                    type: object
+                    properties:
+                      key:
+                        type: string
+                        description: The key to retrieve the permanent link data.
+            400:
+              $ref: '#/components/responses/400'
+            401:
+              $ref: '#/components/responses/401'
+            422:
+              $ref: '#/components/responses/422'
+            500:
+              $ref: '#/components/responses/500'
+        """
+        key_type = current_app.config["PERMALINK_KEY_TYPE"]
+        try:
+            item = self.add_model_schema.load(request.json)
+            state = item["state"]
+            key = CreateDashboardPermalinkCommand(
+                actor=g.user, id_or_slug=pk, state=state, key_type=key_type,
+            ).run()
+            http_origin = request.headers.environ["HTTP_ORIGIN"]
```

Comment:
Do we need to worry about these two cases for [HTTP Origin](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Origin#description)?

> The Origin header value may be null in a number of cases, including (non-exhaustively):
> - Redirects across origins.
> - iframes with a sandbox attribute that doesn't contain the value allow-same-origin.

Reply:
Fixed (this came up while writing integration tests)